src/data/int/log.lean

```diff
@@ -0,0 +1,245 @@
+if_pos hr
+
+lemma log_of_right_le_one (b : ℕ) {r : R} (hr : r ≤ 1) : log b r = -nat.clog b ⌈r⁻¹⌉₊ :=
+begin
+  obtain rfl | hr := hr.eq_or_lt,
+  { rw [log, if_pos hr, inv_one, nat.ceil_one, nat.floor_one, nat.log_one_right, nat.clog_one_right,
+        int.coe_nat_zero, neg_zero], },
+  { exact if_neg hr.not_le }
+end
+
+@[simp, norm_cast] lemma log_nat_cast (b : ℕ) (n : ℕ) : log b (n : R) = nat.log b n :=
+begin
+  cases n,
+  { simp [log_of_right_le_one _ _, nat.log_zero_right] },
+  { have : 1 ≤ (n.succ : R) := by simp,
+    simp [log_of_one_le_right _ this, ←nat.cast_succ] }
+end
+
+lemma log_of_left_le_one {b : ℕ} (hb : b ≤ 1) (n : R) : log b n = 0 :=
+begin
+  cases le_total 1 n,
+  { rw [log_of_one_le_right _ h, nat.log_of_left_le_one hb, int.coe_nat_zero] },
+  { rw [log_of_right_le_one _ h, nat.clog_of_left_le_one hb, int.coe_nat_zero, neg_zero] },
+end
+
+lemma log_of_right_le_zero {n : R} (hn : n ≤ 0) (b : ℕ) : log b n = 0 :=
+by rw [log_of_right_le_one _ (hn.trans zero_le_one),
+    nat.clog_of_right_le_one ((nat.ceil_eq_zero.mpr $ inv_nonpos.2 hn).trans_le zero_le_one),
+    int.coe_nat_zero, neg_zero]
+
+lemma zpow_log_le_self {b : ℕ} {r : R} (hn : 1 < b) (hr : 0 < r) :
+  (b : R) ^ log b r ≤ r :=
+begin
+  cases le_total 1 r with hr1 hr1,
+  { rw log_of_one_le_right _ hr1,
+    refine le_trans _ (nat.floor_le hr.le),
+    rw [zpow_coe_nat, ←nat.cast_pow, nat.cast_le],
+    exact nat.pow_log_le_self hn (nat.floor_pos.mpr hr1) },
+  { rw [log_of_right_le_one _ hr1, zpow_neg, zpow_coe_nat, ← nat.cast_pow],
+    apply inv_le_of_inv_le hr,
+    refine (nat.le_ceil _).trans (nat.cast_le.2 _),
+    exact nat.le_pow_clog hn _ },
+end
+
+lemma lt_zpow_succ_log_self (b : ℕ) (r : R) (hn : 1 < b) (hr : 0 < r) :
+  r < (b : R) ^ (log b r + 1) :=
+begin
+  cases le_or_lt 1 r with hr1 hr1,
+  { rw log_of_one_le_right _ hr1,
+    rw [int.coe_nat_add_one_out, zpow_coe_nat, ←nat.cast_pow],
+    apply nat.lt_of_floor_lt,
+    exact nat.lt_pow_succ_log_self hn _, },
+  { rw log_of_right_le_one _ hr1.le,
+    have hcri : 1 < r⁻¹ := one_lt_inv hr hr1,
+    have : 1 ≤ nat.clog b ⌈r⁻¹⌉₊ :=
+      nat.succ_le_of_lt (nat.clog_pos hn $ nat.one_lt_cast.1 $ hcri.trans_le (nat.le_ceil _)),
+    rw [neg_add_eq_sub, ←neg_sub, ←int.coe_nat_one, ← int.coe_nat_sub this,
+      zpow_neg, zpow_coe_nat, lt_inv hr (pow_pos (nat.cast_pos.mpr $ zero_lt_one.trans hn) _),
+      ←nat.cast_pow],
+    refine nat.lt_ceil.1 _,
+    exact (nat.pow_pred_clog_lt_self hn $ nat.one_lt_cast.1 $ hcri.trans_le $ nat.le_ceil _), }
+end
+
+@[simp] lemma log_zero_right (b : ℕ) : log b (0 : R) = 0 :=
+log_of_right_le_zero le_rfl _
+
+@[simp] lemma log_one_right (b : ℕ) : log b (1 : R) = 0 :=
+by rw [log_of_one_le_right _ le_rfl, nat.floor_one, nat.log_one_right, int.coe_nat_zero]
+
+/-- `zpow b` and `int.log b` (almost) form a Galois connection. -/
+lemma zpow_le_iff_le_log {b : ℕ} (hb : 1 < b) {x : ℤ} {r : R} (hr : 0 < r) :
+  (b : R) ^ x ≤ r ↔ x ≤ log b r :=
```

Comment:
Can you also prove `r < b ^ x ↔ log b r < x`? and same for `clog`. I suspect it might be easier to prove directly than through this current lemma (because the proof is so long), so you might want to swap them.

Reply:
Do we have that lemma for `nat.log`?

Reply:
> I suspect it might be easier to prove directly than through this current lemma (because the proof is so long), so you might want to swap them.

I couldn't see an obvious way to prove one in terms of the other. I pushed the stupid proof that is just the `le` proof with `le`s swapped with `lt`s. Can you try to golf one in terms of the other?